clients/src/main/java/org/apache/kafka/clients/producer/internals/TransactionManager.java

```diff
@@ -239,12 +243,14 @@ private boolean isTransitionValid(State source, State target) {
 
     // We use the priority to determine the order in which requests need to be sent out. For instance, if we have
     // a pending FindCoordinator request, that must always go first. Next, If we need a producer id, that must go second.
-    // The endTxn request must always go last.
+    // The endTxn request must always go last, unless we are bumping the epoch (a special case of InitProducerId)as
+    // part of ending the transaction.
     private enum Priority {
         FIND_COORDINATOR(0),
         INIT_PRODUCER_ID(1),
         ADD_PARTITIONS_OR_OFFSETS(2),
-        END_TXN(3);
+        END_TXN(3),
+        EPOCH_BUMP(4);
```

Comment:
Once we have decided to bump the epoch, is there any harm in doing so right away? In other words, do we benefit by ending the transaction first?

Reply:
I initially thought that doing the state tracking would be hard if we just bumped right away without aborting. But actually it should be fine, since InitProducerId already aborts any ongoing transactions for that producer. It will have to retry at least once, because on the first call it gets a CONCURREN_TRANSACTIONS error, but that error is already handled. So this suggestion is actually what I went with.